src/braket/experimental/autoqasm/errors.py

```diff
@@ -24,12 +22,28 @@ class UnknownQubitCountError(AutoQasmError):
     """Missing declaration for the number of qubits."""
 
     def __init__(self):
-        self.message = f"""Unspecified number of qubits.
+        self.message = """Unspecified number of qubits.
+
+Specify the number of qubits used by your program by supplying the \
+`num_qubits` argument to `autoqasm.function`. For example:
 
-Please declare the total number of qubits for your program. \
-You can do that by calling your AutoQASM program with the \
-{ProgramOptions.NUM_QUBITS.value} keyword argument. \
+    @autoqasm.function(num_qubits=5)
+    def my_autoqasm_function():
+        ...
 """
 
     def __str__(self):
         return self.message
+
+
+class InconsistentNumQubits(AutoQasmError):
+    """Num qubits supplied to main function does not match subroutine."""
```

Comment:
nit: "main function" and "subroutine" are not really terms we have exposed to the user. But I think it's understandable regardless - probably this is fine.

Reply:
I know, writing this made me feel like we do need to make the distinction. because I couldn't figure out how to explain the difference between the one you call and the one that one calls in a clear way :/ Open to ideas

Reply:
Yeah. It's difficult because customers may feel like they need to specify the number of qubits each time they use the @aq.function decorator, which means that they need to keep them all in sync (or have errors thrown) and when they want to change their algorithm to use more qubits, they would need to revisit every function they created.

Reply:
`num_qubits` should only be specified at the top level AutoQASM function. They should not be specified by nested subroutines, so there's no need to sync them up. However, we have no such distinguishing main/subroutine feature at the moment, so I can't be that precise yet.

Correct:
```
@aq.function
def nested():
    ...

@aq.function(num_qubits=5)
def main():
    nested()
```

Incorrect:
```
@aq.function(num_qubits=10)
def nested():
    ...

@aq.function(num_qubits=5)
def main():
    nested()
```

Not preferred, but not a hard failure:
```
@aq.function(num_qubits=5)
def nested():
    ...

@aq.function(num_qubits=5)
def main():
    nested()
```

Do you think the last case should be a hard failure?

Reply:
It would block the functionality you describe. I'm not convinced it's an important use case. It feels analogous to a package (lets say C or Rust) that has some defined executables, and a bunch of library code. The library code can be called from the executables. It would be odd and awkward to try to use one executable from another executable.

I did leave it for flexibility (allowing the nested code to be called directly) and it like the right approach until we decide to split `aq.main` and `aq.subroutine`